phonenumberutil.go

```diff
@@ -549,25 +556,127 @@ var (
 	// Note: Synchronization, though only needed for the Android version
 	// of the library, is used in all versions for consistency.
 	countryCodeToNonGeographicalMetadataMap = make(map[int]*PhoneMetadata)
+	ccToNGMMutex                            sync.RWMutex
 
 	// A cache for frequently used region-specific regular expressions.
 	// The initial capacity is set to 100 as this seems to be an optimal
 	// value for Android, based on performance measurements.
-	regexCache = make(map[string]*regexp.Regexp)
+	regexCache    = make(map[string]*regexp.Regexp)
+	regCacheMutex sync.RWMutex
 
 	// The set of regions the library supports.
 	// There are roughly 240 of them and we set the initial capacity of
 	// the HashSet to 320 to offer a load factor of roughly 0.75.
 	supportedRegions = make(map[string]struct{})
+	sRMutex          sync.RWMutex
 
 	// The set of county calling codes that ma`p to the non-geo entity
 	// region ("001"). This set currently contains < 12 elements so the
 	// default capacity of 16 (load factor=0.75) is fine.
 	countryCodesForNonGeographicalRegion = make(map[int]struct{})
+	ccForNGRMutex                        sync.RWMutex
 )
 
 var ErrEmptyMetadata = errors.New("empty metadata")
 
+func readFromRegexCache(key string) (*regexp.Regexp, bool) {
+	regCacheMutex.RLock()
+	v, ok := regexCache[key]
+	regCacheMutex.RUnlock()
+	runtime.Gosched()
```

Comment:
Could we remove all the calls to Gosched? I don't think it's a library's job to try to directly interact with the scheduler if it doesn't specifically say that it's going to.